### PR TITLE
Fix to #16092 - Query: Simplify case blocks in SQL tree

### LINKED ISSUE #16092: Query: Simplify case blocks in SQL tree
Currently we translate CompareTo naively, using case statements with 3 options, however those can be significantly optimized for cases where CompareTo itself is compared to 1, 0 -1

foo.CompareTo(bar) == 0 -> foo == bar

example test:
Double_order_by_on_string_compare

example current sql:

```sql
SELECT [w].[Id], [w].[AmmunitionType], [w].[IsAutomatic], [w].[Name], [w].[OwnerFullName], [w].[SynergyWithId]
FROM [Weapons] AS [w]
ORDER BY CASE
    WHEN (CASE
        WHEN ([w].[Name] = N'Marcus'' Lancer') AND [w].[Name] IS NOT NULL THEN 0
        WHEN [w].[Name] > N'Marcus'' Lancer' THEN 1
        WHEN [w].[Name] < N'Marcus'' Lancer' THEN -1
    END = 0) AND CASE
        WHEN ([w].[Name] = N'Marcus'' Lancer') AND [w].[Name] IS NOT NULL THEN 0
        WHEN [w].[Name] > N'Marcus'' Lancer' THEN 1
        WHEN [w].[Name] < N'Marcus'' Lancer' THEN -1
    END IS NOT NULL THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END, [w].[Id]
```

example improved sql:
```sql
SELECT [w].[Id], [w].[AmmunitionType], [w].[IsAutomatic], [w].[Name], [w].[OwnerFullName], [w].[SynergyWithId]
FROM [Weapons] AS [w]
ORDER BY CASE
    WHEN [w].[Name] = N'Marcus'' Lancer'
    THEN CAST(1 AS bit) ELSE CAST(0 AS bit)
END, [w].[Id]
```

### PR DESCRIPTION
**Pull request by maumar**
Adding optimization to during post processing (null semantics)
Trying to match CASE block that corresponds to CompareTo translation. If that case block is compared to 0, 1 or -1 we can simplify it to simple comparison.
Also added generic CASE block optimization, when constant is compared to CASE block, and that constant is one of the results

Fixes #16092

### CODE REVIEW

File: src/EFCore.Relational/Query/SqlNullabilityProcessor.cs [hunk 1]
```diff
@@ -751,19 +751,52 @@ protected virtual SqlExpression VisitLike([NotNull] LikeExpression likeExpressio
...
         /// <param name="allowOptimizedExpansion"> A bool value indicating if optimized expansion which considers null value as false value is allowed. </param>
         /// <param name="nullable"> A bool value indicating whether the sql expression is nullable. </param>
         /// <returns> An optimized sql expression. </returns>
-
         protected virtual SqlExpression VisitSqlBinary(
             [NotNull] SqlBinaryExpression sqlBinaryExpression, bool allowOptimizedExpansion, out bool nullable)
         {
             Check.NotNull(sqlBinaryExpression, nameof(sqlBinaryExpression));
 
+            // we need to do this before we visit left/right
```
Comment (maumar) [inline src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:767]: could put it in a separate visitor that runs before null semantics. slower perf (so prolly bad idea), but makes the design cleaner. thoughts?

**Comment by maumar** [comment]
@smitpatel it's not rly part of null semantics. it needs to happen before null semantics. I can put it in a separate visitor, as mentioned in a comment

**Comment by smitpatel** [comment]
Does it need to know parameter values?

**Comment by maumar** [comment]
yes, I will add some tests

File: src/EFCore.Relational/Query/Internal/CaseWhenFlatteningExpressionVisitor.cs [hunk 2]
```diff
@@ -56,7 +56,164 @@ protected override Expression VisitExtension(Expression extensionExpression)
                     nestedCaseExpression.ElseResult));
             }
 
+            if (extensionExpression is SqlBinaryExpression sqlBinaryExpression)
```
Comment (maumar) [inline src/EFCore.Relational/Query/Internal/CaseWhenFlatteningExpressionVisitor.cs:59]: rename the class to something like CaseSimplifyingExpressionVisitor?

**Comment by maumar** [comment]
updated

File: src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+                // generic CASE statement comparison optimization:
+                // (CASE
+                //  WHEN condition1 THEN result1
+                //  WHEN condition2 THEN result2
+                //  WHEN ...
+                //  WHEN conditionN THEN resultN) == result1 -> condition1
+                if (sqlBinaryExpression.OperatorType == ExpressionType.Equal
+                    && sqlConstantComponent != null
+                    && sqlConstantComponent.Value != null
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:72]: Why this value has to be null? Is there anything wrong in comparing null with null?

Comment (maumar) [reply src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:72]: when case statement doesn't match anything null is also returned. We would miss this case if we matched here. Maybe it never/rarely happens but I added the condition just to be on the safe side

Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:72]: A long optimization if you are interested in doing,
If constant value is null then,
Main condition is when all tests from when clauses are false (negate & join with AND)
If ElseResult is present then add ElseResult == null with AND
If a when clause returns null result then add that when clause test with OR with all of above.

Bit complex so don't know if worth having it. Your call.

Comment (maumar) [reply src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:72]: i thought of that, not worth imo

**Comment by maumar** [comment]
ping

File: src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                    && caseComponent != null
+                    && caseComponent.Operand == null)
+                {
+                    var matchingCaseBlock = caseComponent.WhenClauses.FirstOrDefault(wc => sqlConstantComponent.Equals(wc.Result));
+                    if (matchingCaseBlock != null)
+                    {
+                        return matchingCaseBlock.Test;
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:79]: Possibly visit recursively. Also write a regression test for it.

File: src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                    }
+                }
+
+                // CompareTo specific optimizations
+                if (sqlConstantComponent != null
+                    && sqlConstantComponent.Value != null
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:85]: `is int` pattern match will verify that it is non-null.

Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:85]: Just add additional check for compare to that ElseResult is also null.

File: src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+                    && IsCompareTo(caseComponent)
+                    && sqlConstantComponent.Value is int intValue
+                    && (intValue == 0 || intValue == 1 || intValue == -1)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:88]:
```suggestion
                    && (intValue > -2 && intValue < 2)
```
could be faster. :trollface:

File: src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+            }
+
+            return base.VisitExtension(extensionExpression);
+        }
+
+        private bool IsCompareTo(CaseExpression caseExpression)
+        {
+            if (caseExpression != null
+                && caseExpression.Operand == null
+                && caseExpression.WhenClauses.Count == 3
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:109]: No ElseResult.

File: src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                : sqlBinaryExpression.OperatorType switch
+                {
+                    ExpressionType.GreaterThan => ExpressionType.LessThan,
+                    ExpressionType.GreaterThanOrEqual => ExpressionType.LessThanOrEqual,
+                    ExpressionType.LessThan => ExpressionType.GreaterThan,
+                    ExpressionType.LessThanOrEqual => ExpressionType.GreaterThanOrEqual,
+                    _ => sqlBinaryExpression.OperatorType
+                };
+
+            if (operatorType == ExpressionType.NotEqual)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs:156]: use case.

**Comment by maumar** [comment]
@smitpatel updated

**Pull request merged**